### PR TITLE
Consistency check for unvisited notifications with read messages

### PR DESCRIPTION
**Pull request by gmethvin**

### CODE REVIEW

File: shoebox/app/com/keepit/model/UserNotification.scala [hunk 1]
```diff
@@ -64,6 +70,26 @@ class UserNotificationRepoImpl @Inject() (
...
+        |WHERE n.state NOT IN ('$SUBSUMED','$VISITED','$INACTIVE')
+        |  AND n.category = '$MESSAGE'
+        |  AND m.id = n.comment_id
+        |  AND (m.id = r.parent_id OR m.parent = r.parent_id)
+        |  AND r.user_id = n.user_id
+        |  AND r.last_read_id >= n.comment_id;
+      """.stripMargin
+    val rs = s.getPreparedStatement(idQuery).executeQuery()
+    val ids = mutable.ArrayBuffer[Id[UserNotification]]()
+    while (rs.next()) { ids += Id(rs.getLong("id")) }
```
Comment (eishay) [inline shoebox/app/com/keepit/model/UserNotification.scala]: best practice is to do 

``` scala
try {use resultSet} finally {resultSet.close}
```

Comment (eishay) [inline shoebox/app/com/keepit/model/UserNotification.scala]: in `SocialConnectionRepoImpl` we use something like 

``` scala
    val q = StaticQuery.query[(Long, Long), Long](sql)
    val res: Seq[Long] = q.list(id.id, id.id)
    res map {id => Id[User](id)} toSet
```

it does basically the same thing, manages the string replacements and closing resources, we can use either.

File: shoebox/test/com/keepit/realtime/NotificationConsistencyCheckerTest.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          (user, nUri, comment, notif)
+        }
+        checker.verifyVisited()
+        healthcheck.errors must beEmpty
+        db.readWrite { implicit s =>
+          commentReadRepo.save(
+            CommentRead(userId = user.id.get, uriId = nUri.id.get, lastReadId = comment.id.get, parentId = comment.id))
+        }
+        checker.verifyVisited()
+        healthcheck.errors must not beEmpty
```
Comment (eishay) [inline shoebox/test/com/keepit/realtime/NotificationConsistencyCheckerTest.scala:40]: nice test!

File: shoebox/test/com/keepit/realtime/NotificationConsistencyCheckerTest.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import com.keepit.model._
+import com.keepit.test.{EmptyApplication, DbRepos}
+
+import akka.actor.ActorSystem
+import akka.testkit.TestKit
+import play.api.Play.current
+import play.api.test.Helpers.running
+import play.api.libs.json.Json
+
+class NotificationConsistencyCheckerTest extends TestKit(ActorSystem()) with Specification with DbRepos {
```
Comment (eishay) [inline shoebox/test/com/keepit/realtime/NotificationConsistencyCheckerTest.scala:16]: cool, i would like to learn more about TestKit. can you do a short page in the wiki about how to test with actors?

Comment (gmethvin) [inline shoebox/test/com/keepit/realtime/NotificationConsistencyCheckerTest.scala:16]: Yeah, I can do that.

File: shoebox/app/com/keepit/model/UserNotification.scala (hunk 1 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/model/UserNotification.scala]: Cool, thanks for catching this.

**Pull request merged**